analytics/src/test/java/com/segment/analytics/SegmentIntegrationTest.kt

```diff
@@ -0,0 +1,653 @@
+        val client = mock(Client::class.java)
+        whenever(client.upload())
+            .thenReturn(
+                object : Client.Connection(
+                    mock(
+                        HttpURLConnection::class.java
+                    ),
+                    mock(InputStream::class.java),
+                    mock(OutputStream::class.java)
+                ) {
+                    @Throws(IOException::class)
+                    override fun close() {
+                        throw Client.HTTPException(429, "Too Many Requests", "too many requests")
+                    }
+                })
+        val segmentIntegration = SegmentBuilder()
+            .client(client)
+            .payloadQueue(payloadQueue)
+            .build()
+        for (i in 0 until 4) {
+            payloadQueue.add(TRACK_PAYLOAD_JSON.toByteArray())
+        }
+        segmentIntegration.submitFlush()
+
+        // Verify that messages were not removed from the queue when server returned a 429.
+        assertThat(queueFile.size()).isEqualTo(4)
+        verify(client).upload()
+    }
+
+    @Test
+    @Throws(IOException::class)
+    fun serializationErrorSkipsAddingPayload() {
+        val payloadQueue = mock(PayloadQueue::class.java)
+        val cartographer = mock(Cartographer::class.java)
+        val payload = Builder().event("event").userId("userId").build()
+        val segmentIntegration = SegmentBuilder()
+            .cartographer(cartographer)
+            .payloadQueue(payloadQueue)
+            .build()
+
+        // Serialized json is null.
+        whenever(cartographer.toJson(any(Map::class.java))).thenReturn(null)
+        segmentIntegration.performEnqueue(payload)
+        verify(payloadQueue, never()).add(any<Any>() as ByteArray?)
+
+        // Serialized json is empty.
+        whenever(cartographer.toJson(any(Map::class.java))).thenReturn("")
+        segmentIntegration.performEnqueue(payload)
+        verify(payloadQueue, never()).add(any<Any>() as ByteArray?)
+
+        // Serialized json is too large (> MAX_PAYLOAD_SIZE).
+        val stringBuilder = StringBuilder()
+        for (i in 0 until MAX_PAYLOAD_SIZE + 1) {
+            stringBuilder.append("a")
+        }
+        whenever(cartographer.toJson(any(Map::class.java))).thenReturn(stringBuilder.toString())
+        segmentIntegration.performEnqueue(payload)
+        verify(payloadQueue, never()).add(any<Any>() as ByteArray?)
+    }
+
+    @Test
+    @Throws(IOException::class)
+    fun shutDown() {
+        val payloadQueue = mock(PayloadQueue::class.java)
+        val segmentIntegration = SegmentBuilder().payloadQueue(payloadQueue).build()
+
+        segmentIntegration.shutdown()
+
+        verify(payloadQueue).close()
+    }
+
+    @Test
+    @Throws(IOException::class)
+    fun payloadVisitorReadsOnly475KB() {
+        val payloadWriter = PayloadWriter(
+            mock(BatchPayloadWriter::class.java), Crypto.none()
+        )
+        val bytes =
+            """{
```

Comment:
for readability it would be good if the starting `{` is on a different line. something like
```
"""
{
   ....
}
"""

Reply:
I tried changing the brackets to be in that form but it changes the output of the test and makes it fail? I think this is because of how byte arrays are formed?

Reply:
oh interesting. i guess ignore for now